src/wallet/connection.ts

```diff
@@ -0,0 +1,123 @@
+ */
+
+import { EventEmitter } from 'events';
+import networkInstance from '../network';
+import { DEFAULT_SERVERS } from '../constants';
+import Network from '../models/network';
+import version from '../version';
+import helpers from '../helpers';
+import wallet from '../wallet';
+import WS from './websocket';
+import config from '../config';
+
+export enum ConnectionState {
+  CLOSED = 0,
+  CONNECTING = 1,
+  CONNECTED = 2,
+}
+
+/**
+ * This is a Connection that may be shared by one or more wallets.
+ *
+ * It has the following states:
+ * - CLOSED: When it is disconnected from the server.
+ * - CONNECTING: When it is connecting to the server.
+ * - CONNECTED: When it is connected.
+ *
+ * You can subscribe for the following events:
+ * - state: Fired when the state of the Wallet changes.
+ * - wallet-update: Fired when a new wallet message arrive from the websocket.
+ **/
+class Connection extends EventEmitter {
+  // the network to connect to, 'testnet' or 'mainnet'
+  private network: Network;
+  private state: ConnectionState;
+  private websocket: WS;
+  private walletId: string;
+
+  /*
+   * network {String} 'testnet' or 'mainnet'
+   */
+  constructor({
+    network = new Network('mainnet'),
+    walletId = '',
+    connectionTimeout = null,
+  } = {}) {
+    super();
+
+    if (!network) {
+      throw Error('You must explicitly provide the network.');
+    }
+
+    if (!walletId || !walletId.length) {
+      throw Error('You must explicitly provide the walletId.');
+    }
+
+    networkInstance.setNetwork(network.name);
+    this.network = network;
+
+    this.walletId = walletId;
+
+    this.state = ConnectionState.CLOSED;
+
+    this.onConnectionChange = this.onConnectionChange.bind(this);
+
+    const wsOptions = {
+      wsURL: config.getWalletServiceBaseWsUrl(this.network),
+      walletId: this.walletId,
+    };
+
+    if (connectionTimeout) {
+      wsOptions['connectionTimeout'] = connectionTimeout;
+    }
+    this.websocket = new WS(wsOptions);
+  }
+
+  /**
+   * Called when the connection to the websocket changes.
+   * It is also called if the network is down.
+   **/
+  onConnectionChange(value: boolean) {
+    if (value) {
+      this.setState(ConnectionState.CONNECTED);
+    } else {
+      this.setState(ConnectionState.CONNECTING);
+    }
+  }
+
+  /**
+   * Update class state
+   */
+  setState(state: ConnectionState) {
+    this.state = state;
+    this.emit('state', state);
+  }
+
+  /**
+   * Connect to the server and start emitting events.
+   **/
+  start() {
+    this.websocket.on('is_online', this.onConnectionChange);
+    this.websocket.on('new-tx', (payload) => this.emit('new-tx', payload.data));
+    this.websocket.on('update-tx', (payload) => this.emit('update-tx', payload.data));
+
+    this.setState(ConnectionState.CONNECTING);
+    this.websocket.setup();
+  }
+
+  /**
+   * Close the connections and stop emitting events.
+   **/
+  stop() {
```

Comment:
Why you don't end websocket connection on stop?

Maybe we should have a parent class that both Connection classes extend and change

Reply:
This was probably a mistake, can you check again?